apps/whale-api/src/module.api/consortium.service.ts

```diff
@@ -1,17 +1,130 @@
     protected readonly rpcClient: JsonRpcClient,
-    private readonly defidCache: DeFiDCache
+    private readonly defidCache: DeFiDCache,
+    protected readonly cache: SemaphoreCache,
+    protected readonly transactionMapper: TransactionMapper
   ) {}
 
+  private formatTransactionResponse (tx: AccountHistory, members: ConsortiumMember[]): Transaction {
+    return {
+      type: tx.type === 'MintToken' ? 'Mint' : 'Burn',
+      member: members.find(m => m.address === tx.owner)?.name ?? '',
+      tokenAmounts: tx.amounts.map((a: any) => {
+        const splits = a.split('@')
+        return { token: splits[1], amount: splits[0] }
+      }),
+      txId: tx.txid,
+      address: tx.owner,
+      block: tx.blockHeight
+    }
+  }
+
+  async getTransactionHistory (pageIndex: number, limit: number, search: string): Promise<ConsortiumTransactionResponse> {
+    const attrs = (await this.rpcClient.masternode.getGov('ATTRIBUTES')).ATTRIBUTES
+    const members: ConsortiumMember[] = []
+    const searching: boolean = search !== ''
+    const keys: string[] = Object.keys(attrs)
+    const values: object[] = Object.values(attrs)
+    const membersKeyRegex: RegExp = /^v0\/consortium\/\d+\/members$/
+    let totalTxCount: number = 0
+    let searchFound: boolean = false
+
+    keys.forEach((key: string, i: number) => {
+      if (membersKeyRegex.exec(key) !== null) {
+        const membersPerToken = values[i]
+        const memberIds: string[] = Object.keys(membersPerToken)
+        let memberDetails: Array<{ ownerAddress: string, name: string }> = Object.values(membersPerToken)
+
+        if (searching) {
+          const foundMembers = memberDetails.filter(m => m.ownerAddress === search || m.name.toLowerCase().includes(search))
+          if (foundMembers.length > 0) {
+            memberDetails = foundMembers
+            searchFound = true
+          }
+        }
+
+        for (let j = 0; j < memberDetails.length; j++) {
+          const memberId = memberIds[j]
+          if (members.find(m => m.id === memberId) === undefined) {
+            members.push({
+              id: memberId,
+              name: memberDetails[j].name,
+              address: memberDetails[j].ownerAddress
+            })
+          }
+        }
+      }
+    })
+
+    if (searching && !searchFound) {
+      const foundTx = await this.transactionMapper.get(search)
+      if (foundTx !== undefined) {
+        const transactionsOnBlock = await this.rpcClient.account.listAccountHistory('all', { maxBlockHeight: foundTx.block.height, depth: 0 })
+        const transaction = transactionsOnBlock.find(tx => tx.txid === foundTx.txid)
+
+        if (transaction === undefined) {
+          return {
+            total: 0,
+            transactions: []
+          }
+        }
+
+        return {
+          total: 1,
+          transactions: [this.formatTransactionResponse(transaction, members)]
+        }
+      }
+
+      return {
+        total: 0,
+        transactions: []
+      }
```

Comment:
Think this can be optimised.

Reply:
Did not find much to refactor here in this block, but have done minor changes. Please let me know if you see a way to improve this further. 🙏🏼